recipes/openai_textcat.py

```diff
@@ -0,0 +1,287 @@
+from recipes.openai import read_prompt_examples
+
+CSS_FILE_PATH = Path(__file__).parent / "style.css"
+DEFAULT_PROMPT_PATH = (
+    Path(__file__).parent.parent / "templates" / "textcat_prompt.jinja2"
+)
+HTML_TEMPLATE = """
+<div class="cleaned">
+  {{ #label }}
+    <centering>
+    <h2>OpenAI ChatGPT says: {{ meta.answer }}</h2>
+    </centering>
+  {{ /label }}
+  <details>
+    <summary>Show the prompt for OpenAI</summary>
+    <pre>{{openai.prompt}}</pre>
+  </details>
+  <details>
+    <summary>Show the response from OpenAI</summary>
+    <pre>{{openai.response}}</pre>
+  </details>
+</div>
+"""
+
+# Set up openai
+load_dotenv()  # take environment variables from .env.
+
+
+@dataclass
+class TextCatPromptExample(PromptExample):
+    """An example to be passed into an OpenAI TextCat prompt."""
+
+    text: str
+    answer: str
+    reason: str
+
+    @classmethod
+    def from_prodigy(cls, example: Dict, labels: Iterable[str]) -> "PromptExample":
+        """Create a prompt example from Prodigy's format."""
+        if "text" not in example:
+            raise ValueError("Cannot make PromptExample without text")
+
+        full_text = example["text"]
+        reason = example["meta"].get("reason")
+        if len(labels) == 1:
+            answer = "accept" if example.get("answer") else "reject"
+        else:
+            answer = example.get("accept")
+        return cls(text=full_text, answer=answer, reason=reason)
+
+
+class TextCatOpenAISuggester(OpenAISuggester):
+    def parse_response(self, example: Dict, response: str, nlp: Language) -> Dict:
+
+        # Add meta to display OpenAI 'reason'
+        if "meta" not in example:
+            example["meta"] = {}
+
+        example = (
+            self._parse_binary(example, response)
+            if len(self.labels) == 1
+            else self._parse_multi(example, response)
+        )
+        return example
+
+    def _parse_output(self, text: str) -> Dict[str, str]:
+        """Parse the actual text response from OpenAI."""
+        output = {}
+        if text and any(k in text for k in ("answer", "reason")):
+            for line in text.strip().split("\n"):
+                if line and ":" in line:
+                    key, value = line.split(":", 1)
+                    output[key.strip().lower()] = value.strip()
+        else:
+            output = {"answer": "", "reason": ""}
+        return output
+
+    def _parse_binary(self, example: Dict, response: str) -> Dict:
+        """Parse binary TextCat where the 'answer' key means it's a positive class."""
+        output = self._parse_output(response)
+        example["answer"] = output["answer"].lower() == "accept"
+        example["meta"] = {
+            "answer": output["answer"].upper(),
+            "reason": output["reason"],
+        }
+        example["label"] = self.labels[0]
+        return example
+
+    def _parse_multi(self, example: Dict, response: str) -> Dict:
```

Comment:
I have a file called `one-off.jsonl` with this example in it:

```
{"text":"@Tesco just for completeness ordered arrived just now... over an hour late https:\/\/t.co\/NGydnyAm6Q"}
```

Next, I tell OpenAI to annotate it. 

```
python -m prodigy textcat.openai.fetch one-off.jsonl templates/textcat_prompt.jinja2 one-off-annot-v2.jsonl --label "website purchase,complaint,delivery related,compliment" --batch-size 10 --examples-path examples.yaml --verbose -F recipes/openai_textcat.py
```

Then, when I look in the annotated json, these are the contents:

```
{
  "text": "@Tesco just for completeness ordered arrived just now... over an hour late https://t.co/NGydnyAm6Q",
  "openai": {
    "prompt": "\n    From the text below, classify if it belongs to the following labels: website purchase, complaint, delivery related, compliment\n    \n    The task is non-exclusive, so you can provide more than one label as long as\n    they're comma-delimited. For example: Label1, Label2, Label3.\n    \n\n\nAnswer in the following format:\n\nanswer: <string>\nreason: <string>\n\nText:\n\"\"\"\n@Tesco just for completeness ordered arrived just now... over an hour late https://t.co/NGydnyAm6Q\n\"\"\"\n",
    "response": "\nAnswer: Complaint, Delivery related\nReason: The text contains a mention of Tesco, which is a website purchase, as well as a mention of the delivery being late, which is a complaint and delivery related."
  },
  "tokens": [...],
  "meta": {
    "reason": ""
  },
  "options": [
    {
      "id": "website purchase",
      "text": "website purchase"
    },
    {
      "id": "complaint",
      "text": "complaint"
    },
    {
      "id": "delivery related",
      "text": "delivery related"
    },
    {
      "id": "compliment",
      "text": "compliment"
    }
  ],
  "answer": "accept",
  "accept": [],
  "_input_hash": 1284924640,
  "_task_hash": 1095851038
}
```

The answer is empty, and so is the reason. I've seen this happen to a bunch of my examples, so I think something may be up with the parsing.

Reply:
Yeah, I just checked the tests and it seems like the `test_parser_response_*` tests are failing. 

That suggests to me that we also want to make sure that the CI actually runs on this PR now. @ljvmiranda921 I just added you as a writer to this repo, could you make a new PR that isn't forked from your repo? That should allow Prodigy to download in the CI and the tests to run.